driver/driver.go

```diff
@@ -0,0 +1,102 @@
+package driver
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/brimsec/zq/proc"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zqd/api"
+)
+
+type Driver interface {
+	Warn(msg string) error
+	Write(channelID int, batch zbuf.Batch) error
+	ChannelEnd(channelID int, stats api.ScannerStats) error
+	Stats(api.ScannerStats) error
+}
+
+func Run(out *proc.MuxOutput, d Driver, statsInterval time.Duration) error {
```

Comment:
nit: This would be more work, but it would be nice if when `statsInterval` is set to 0 a Noop Ticker is used. I think this would make clearer a lot of places in the code where you set the interval to an arbitrarily long interval to serve as functionally the same thing.

Reply:
Yes, i didn't like passing this arbitrary interval around. I'll change so that 0 means no tick.

Reply:
Actually we need a ticker to pass to `Pull()`, we can't not tick. I've changed instead so that if `0` is passed, we switch to an (arbitrary) value of 10 seconds. 
This is now the same behavior as before this PR. This can probably be improved (in a separate PR).

Reply:
Maybe pass a `<-chan time.Time` instead of a `time.Duration`? Then callers that don't want a `time.Ticker` can pass nil, and those that do can create one and pass its channel.

Reply:
@nwt That seems like a good approach but could you explain what the benefit of passing `nil` as a zero value vs `0` would be? Happy to do it but hoping to learn something here...

Reply:
It just strikes me as more intuitive. It's clear that a nil `statsTicks <-chan time.Time` never ticks, but the meaning of a zero `statsInterval time.Duration` isn't obvious.

Reply:
It's obvious that a nil `<- chan time.Time` never ticks, but upon getting such a nil the `Run()` func would create a `time.NewTicker(10 * time.Second)` under the hood, and use those ticks... which is why  `nil` and `0` seem equally non-obvious to me. (This is all just pre-existing behavior, which we should probably change, but I'm not trying to do that as part of this PR).

Hope this clarifies things. Either way, I can change to this if you'd like; just let me know.

Reply:
I was assuming `driver.Run` would always use `statsTicks` directly and would never create its own `timer.Ticker`.

Reply:
And for the record, I'm fine with either `time.Duration` or `<-chan time.Time`.

Reply:
Passing a nil ticker to `(*MuxOutput).Pull(..)` would be a change to the existing behavior of zq and tests: https://github.com/brimsec/zq/blob/d7c70735c855523588207db22a286f6bbde380bb/driver/run.go#L55

(Not saying this change shouldn't be made! but not here, especially as it breaks some tests)

Reply:
> And for the record, I'm fine with either time.Duration or <-chan time.Time.

@nwt 👍. I filed #554 to revisit passing a nil ticker to `(*MuxOutput).Pull(..)`.